text/1012-range-types.rst

```diff
@@ -0,0 +1,277 @@
+
+The base scalar types that ``range`` and ``multirange`` can work with:
+* ``anypoint`` - any scalar that has "points" which can be used to construct
+  a range (like ``int64`` and ``float64``, but not, say, ``json`` and ``str``)
+* ``anydiscrete extending anypoint`` - any scalar that conceptually
+  represents discrete values (like ``int64`` or ``date``)
+* ``anycontiguous extending anypoint`` - any scalar that conceptually
+  represents contiguous values (like ``float64`` or ``datetime``)
+
+The interval types themselves would be defined similar to arrays:
+* ``range<anypoint>`` represents intervals over some concrete ``anypoint``
+  sub-type
+* ``multirange<anypoint>`` represents a set of non-overlapping ``range``
+  values over the corresponding concrete ``anypoint`` sub-type
+
+
+Boundaries
+----------
+
+A ``range`` value should have an upper and lower bound and an indicator for
+each of them whether they are included in the values the ``range`` represents.
+
+It should be possible to omit one or both of the boundaries to represent an
+open-ended interval. The ``{}`` will be used to represent the omitted
+boundary. It should be an error to omit the boundary, but indicate that it is
+inclusive.
+
+
+Constructors
+------------
+
+We will use a function constructor for creating new ``range`` values::
+
+  function range(
+      lower: optional anypoint = {},
+      upper: optional anypoint = {},
+      named only inclow: bool = true,
+      named only inchigh: bool = false
+  ) -> range<anypoint>
+
+The default values are chosen so as to minimize the amount of keystrokes
+required to define ranges for common use-cases such as for iterators.
+
+The ``inclow`` and ``inchigh`` arguments are ``named only`` because when their
+default values need to be overridden it is entirely plausible that only one of
+them needs to be specified and thus using positional arguments would be
+awkward.
+
+A ``multirange`` could be constructed using a function that takes a set of
+basic ``range`` values::
+
+  function multirange(
+      ranges: set of range<anypoint>
+  ) -> multirange<anypoint>
+
+The constructor must normalize the ranges to be non-overlapping and order them
+based on boundaries. This will ensure that when a ``multirange`` is unpacked
+into its component ranges, the order of the ``range`` values would be based on
+their boundaries rather than arbitrary.
+
+
+Accessing range components
+--------------------------
+
+Several functions will be used to access ``range`` components::
+
+  function get_range_upper(range<anypoint>) -> optional anypoint
+  function get_range_lower(range<anypoint>) -> optional anypoint
+  function is_inclusive_lower(range<anypoint>) -> bool
+  function is_inclusive_upper(range<anypoint>) -> bool
+
+
+Accessing multirange components
+-------------------------------
+
+There will be a function for extracting sub-ranges from a multirange::
+
+  function get_subranges(multirange<anypoint>) -> set of range<anypoint>
```

Comment:
Maybe

```suggestion
  function multirange_unpack(multirange<anypoint>) -> set of range<anypoint>
```

?

Reply:
This is very long name... maybe `to_range(multirange<anypoint>)` instead? It's a conversion function of sorts.

Reply:
`to_range` implies that a single range would be returned, `to_ranges` would be better.

Reply:
Fair enough.